File: csharp/binary-search/Iteration2/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        var start = 0;
        var end = input.Length - 1;

        while (start <= end)
        {
            var middle = (start + end) / 2;

            if (input[middle] == value) return middle;
            else if (value > input[middle]) start = middle + 1;
            else end = middle - 1;
        }

        return -1;
    }
}

File: c/binary-search/Iteration1/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
    if (length == 0) return NULL;
    
    int start = 0;
    int end = length - 1;

    while (start <= end) {
        size_t middle = ((start + end) / 2);
        const int * focus = &arr[middle];

        if (value < *focus) {
            end = middle - 1;
        } else if (value > *focus) {
            start = middle + 1;
        } else {
            return focus;
        }
    }

    return NULL;
}
